File: zivid_samples/src/sample_infield_correction.cpp
#include <ctime>
#include <rclcpp/rclcpp.hpp>
#include <std_srvs/srv/trigger.hpp>
#include <stdexcept>
#include <zivid_interfaces/srv/infield_correction_capture.hpp>
#include <zivid_interfaces/srv/infield_correction_compute.hpp>
#include <zivid_interfaces/srv/infield_correction_read.hpp>

static const auto read_only_parameter =
  rcl_interfaces::msg::ParameterDescriptor{}.set__read_only(true);

[[noreturn]] void fatal_error(const rclcpp::Logger & logger, const std::string & message)
{
  RCLCPP_ERROR_STREAM(logger, message);
  throw std::runtime_error(message);
}

template <typename List, typename ToStringFunc>
std::string join_list_to_string(const List & list, ToStringFunc && to_string_func)
{
  return list.empty()
           ? std::string()
           : std::accumulate(
               std::next(std::begin(list)), std::end(list), to_string_func(*std::begin(list)),
               [&](const std::string & str, const auto & entry) {
                 return str + ", " + to_string_func(entry);
               });
}

template <typename Enum>
Enum get_parameter_enum(
  rclcpp::Node & node, const std::string & name, const std::map<std::string, Enum> & name_value_map)
{
  const auto value = node.declare_parameter<std::string>(name, "", read_only_parameter);
  const auto it = name_value_map.find(value);
  if (it == name_value_map.end()) {
    const std::string valid_values =
      join_list_to_string(name_value_map, [](auto && pair) { return pair.first; });
    fatal_error(
      node.get_logger(), "Invalid value for parameter '" + name + "': '" + value +
                           "'. Expected one of: " + valid_values + ".");
  }
  RCLCPP_INFO(node.get_logger(), "%s", ("Got parameter " + name + ": " + value).c_str());
  return it->second;
}

enum class InfieldCorrectionOperation
{
  Verify,
  Correct,
  CorrectAndWrite,
  Read,
  Reset,
};

const std::map<std::string, InfieldCorrectionOperation> infield_correction_operation_map = {
  {"verify", InfieldCorrectionOperation::Verify},
  {"correct", InfieldCorrectionOperation::Correct},
  {"correct_and_write", InfieldCorrectionOperation::CorrectAndWrite},
  {"read", InfieldCorrectionOperation::Read},
  {"reset", InfieldCorrectionOperation::Reset},
};

template <typename ServiceType>
std::shared_ptr<rclcpp::Client<ServiceType>> create_service_client(
  const rclcpp::Node::SharedPtr & node, const std::string & service_name)
{
  auto client = node->create_client<ServiceType>(service_name);
  while (!client->wait_for_service(std::chrono::seconds(3))) {
    if (!rclcpp::ok()) {
      fatal_error(node->get_logger(), "Client interrupted while waiting for service to appear.");
    }
    RCLCPP_INFO(
      node->get_logger(), "Waiting for the %s service to appear...", client->get_service_name());
  }
  return client;
}

bool request_trigger_and_print_response(
  const rclcpp::Node::SharedPtr & node,
  const std::shared_ptr<rclcpp::Client<std_srvs::srv::Trigger>> & client)
{
  auto future = client->async_send_request(std::make_shared<std_srvs::srv::Trigger::Request>());
  if (rclcpp::spin_until_future_complete(node, future) != rclcpp::FutureReturnCode::SUCCESS) {
    fatal_error(node->get_logger(), "Failed to call the infield correction compute service");
  }
  auto result = future.get();
  RCLCPP_INFO(node->get_logger(), "Trigger results (%s):", client->get_service_name());
  RCLCPP_INFO(node->get_logger(), "  Success: %s", result->success ? "true" : "false");
  RCLCPP_INFO(
    node->get_logger(), "  Message: %s",
    result->message.empty() ? "" : (R"(""")" + result->message + R"(""")").c_str());
  return result->success;
}

std::string capture_status_to_string(
  const std::shared_ptr<rclcpp::Node> & node,
  const zivid_interfaces::srv::InfieldCorrectionCapture::Response & response)
{
  using zivid_interfaces::srv::InfieldCorrectionCapture;
  switch (response.status) {
    case InfieldCorrectionCapture::Response::STATUS_NOT_SET:
      return "STATUS_NOT_SET";
    case InfieldCorrectionCapture::Response::STATUS_OK:
      return "STATUS_OK";
    case InfieldCorrectionCapture::Response::STATUS_DETECTION_FAILED:
      return "STATUS_DETECTION_FAILED";
    case InfieldCorrectionCapture::Response::STATUS_INVALID_CAPTURE_METHOD:
      return "STATUS_INVALID_CAPTURE_METHOD";
    case InfieldCorrectionCapture::Response::STATUS_INVALID_ALIGNMENT:
      return "STATUS_INVALID_ALIGNMENT";
    default:
      fatal_error(node->get_logger(), "Invalid status: " + std::to_string(response.status));
  }
}

bool request_capture_and_print_response(
  const rclcpp::Node::SharedPtr & node,
  const std::shared_ptr<rclcpp::Client<zivid_interfaces::srv::InfieldCorrectionCapture>> & client)
{
  auto future = client->async_send_request(
    std::make_shared<zivid_interfaces::srv::InfieldCorrectionCapture::Request>());
  if (rclcpp::spin_until_future_complete(node, future) != rclcpp::FutureReturnCode::SUCCESS) {
    fatal_error(node->get_logger(), "Failed to call the infield correction capture service");
  }
  auto result = future.get();
  RCLCPP_INFO(node->get_logger(), "Trigger results (%s):", client->get_service_name());
  RCLCPP_INFO(node->get_logger(), "  Success: %s", result->success ? "true" : "false");
  RCLCPP_INFO(
    node->get_logger(), "  Message: %s",
    result->message.empty() ? "" : (R"(""")" + result->message + R"(""")").c_str());
  RCLCPP_INFO(node->get_logger(), "  Status: %s", capture_status_to_string(node, *result).c_str());
  RCLCPP_INFO(node->get_logger(), "  Number of captures: %d", result->number_of_captures);

  return result->success;
}

void request_infield_correction_compute_and_print_response(
  const rclcpp::Node::SharedPtr & node,
  const std::shared_ptr<rclcpp::Client<zivid_interfaces::srv::InfieldCorrectionCompute>> & client,
  const std::string & name)
{
  auto future = client->async_send_request(
    std::make_shared<zivid_interfaces::srv::InfieldCorrectionCompute::Request>());
  if (rclcpp::spin_until_future_complete(node, future) != rclcpp::FutureReturnCode::SUCCESS) {
    fatal_error(node->get_logger(), "Failed to call the infield correction " + name + " service");
  }
  auto result = future.get();
  RCLCPP_INFO(
    node->get_logger(), "Infield correction compute results (%s):", client->get_service_name());
  RCLCPP_INFO(node->get_logger(), "  Success: %s", result->success ? "true" : "false");
  RCLCPP_INFO(node->get_logger(), "  Number of captures: %d", result->number_of_captures);

  RCLCPP_INFO(node->get_logger(), "  Trueness errors:");
  for (size_t i = 0; i < result->trueness_errors.size(); i++) {
    RCLCPP_INFO(
      node->get_logger(), "  - Capture %zu: %g %%", i + 1, result->trueness_errors[i] * 100.f);
  }
  RCLCPP_INFO(
    node->get_logger(), "  Average trueness error: %g %%", result->average_trueness_error * 100.f);
  RCLCPP_INFO(
    node->get_logger(), "  Maximum trueness error: %g %%", result->maximum_trueness_error * 100.f);
  RCLCPP_INFO(
    node->get_logger(), "  Dimension accuracy: %g %%", result->dimension_accuracy * 100.f);
  RCLCPP_INFO(node->get_logger(), "  Z min: %g m", result->z_min);
  RCLCPP_INFO(node->get_logger(), "  Z max: %g m", result->z_max);
  RCLCPP_INFO(
    node->get_logger(), "  Message: %s",
    result->message.empty() ? "" : (R"(""")" + result->message + R"(""")").c_str());
}

bool request_read_and_print_response(
  const rclcpp::Node::SharedPtr & node,
  const std::shared_ptr<rclcpp::Client<zivid_interfaces::srv::InfieldCorrectionRead>> & client)
{
  auto future = client->async_send_request(
    std::make_shared<zivid_interfaces::srv::InfieldCorrectionRead::Request>());
  if (rclcpp::spin_until_future_complete(node, future) != rclcpp::FutureReturnCode::SUCCESS) {
    fatal_error(node->get_logger(), "Failed to call the infield correction read service");
  }
  auto result = future.get();
  const std::time_t time = result->camera_correction_timestamp.sec;
  std::ostringstream oss;
  oss << std::put_time(std::localtime(&time), "%c");
  const std::string time_str = oss.str();
  RCLCPP_INFO(node->get_logger(), "Trigger results (%s):", client->get_service_name());
  RCLCPP_INFO(node->get_logger(), "  Success: %s", result->success ? "true" : "false");
  RCLCPP_INFO(
    node->get_logger(), "  Message: %s",
    result->message.empty() ? "" : (R"(""")" + result->message + R"(""")").c_str());
  RCLCPP_INFO(
    node->get_logger(), "  Has camera correction: %s",
    result->has_camera_correction ? "true" : "false");
  RCLCPP_INFO(
    node->get_logger(), "  Camera correction timestamp: %d (%s)",
    result->camera_correction_timestamp.sec, time_str.c_str());

  return result->success;
}

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<rclcpp::Node>("sample_infield_correction");

  const InfieldCorrectionOperation operation =
    get_parameter_enum(*node, "operation", infield_correction_operation_map);

  switch (operation) {
    case InfieldCorrectionOperation::Verify: {
      RCLCPP_INFO(node->get_logger(), "--- Starting infield correction: Verify ---");
      auto start_client =
        create_service_client<std_srvs::srv::Trigger>(node, "infield_correction/start");
      auto capture_client = create_service_client<zivid_interfaces::srv::InfieldCorrectionCapture>(
        node, "infield_correction/capture");
      auto compute_client = create_service_client<zivid_interfaces::srv::InfieldCorrectionCompute>(
        node, "infield_correction/compute");

      if (!request_trigger_and_print_response(node, start_client)) {
        fatal_error(node->get_logger(), "Could not start infield correction, aborting.");
      }
      if (!request_capture_and_print_response(node, capture_client)) {
        fatal_error(
          node->get_logger(), "Could not perform an infield correction capture, aborting.");
      }
      request_infield_correction_compute_and_print_response(node, compute_client, "compute");
    } break;
    case InfieldCorrectionOperation::Correct:
    case InfieldCorrectionOperation::CorrectAndWrite: {
      RCLCPP_INFO(node->get_logger(), "--- Starting infield correction ---");
      auto start_client =
        create_service_client<std_srvs::srv::Trigger>(node, "infield_correction/start");
      if (!request_trigger_and_print_response(node, start_client)) {
        fatal_error(node->get_logger(), "Could not start infield correction, aborting.");
      }

      auto capture_client = create_service_client<zivid_interfaces::srv::InfieldCorrectionCapture>(
        node, "infield_correction/capture");
      auto compute_client = create_service_client<zivid_interfaces::srv::InfieldCorrectionCompute>(
        node, "infield_correction/compute");

      constexpr int wait_seconds = 5;
      constexpr int num_successful_captures_target = 3;
      RCLCPP_INFO_STREAM(
        node->get_logger(), "--- Starting captures, will proceed until " +
                              std::to_string(num_successful_captures_target) +
                              " captures have completed with a detected calibration board ---");

      bool sleep_before_capture = false;
      for (int num_successful_captures = 0;
           num_successful_captures < num_successful_captures_target;) {
        if (sleep_before_capture) {
          RCLCPP_INFO_STREAM(
            node->get_logger(), "--- Waiting for " + std::to_string(wait_seconds) +
                                  " seconds before taking the next capture ---");
          rclcpp::spin_until_future_complete(
            node, std::promise<bool>().get_future(), std::chrono::seconds(wait_seconds));
        } else {
          sleep_before_capture = true;
        }

        const bool success = request_capture_and_print_response(node, capture_client);
        if (success) {
          num_successful_captures += 1;
          // Run compute on the gathered captures so far. This provides useful information about the
          // expected effect of the correction with the current captures.
          request_infield_correction_compute_and_print_response(node, compute_client, "compute");
        }
      }

      RCLCPP_INFO(node->get_logger(), "--- Captures complete ---");

      if (operation == InfieldCorrectionOperation::CorrectAndWrite) {
        RCLCPP_INFO_STREAM(node->get_logger(), "--- Writing correction results to camera ---");
        auto compute_and_write_client =
          create_service_client<zivid_interfaces::srv::InfieldCorrectionCompute>(
            node, "infield_correction/compute_and_write");
        request_infield_correction_compute_and_print_response(
          node, compute_and_write_client, "compute and write");
      }

    } break;
    case InfieldCorrectionOperation::Read: {
      RCLCPP_INFO(node->get_logger(), "--- Starting infield correction: Read ---");
      auto read_client = create_service_client<zivid_interfaces::srv::InfieldCorrectionRead>(
        node, "infield_correction/read");
      request_read_and_print_response(node, read_client);
    } break;
    case InfieldCorrectionOperation::Reset: {
      RCLCPP_INFO(node->get_logger(), "--- Starting infield correction: Reset ---");
      auto reset_client =
        create_service_client<std_srvs::srv::Trigger>(node, "infield_correction/reset");
      request_trigger_and_print_response(node, reset_client);
    } break;
    default: {
      fatal_error(
        node->get_logger(), "Unknown operation: " + std::to_string(static_cast<int>(operation)));
    } break;
  }

  RCLCPP_INFO(node->get_logger(), "Spinning node.. Press Ctrl+C to abort.");
  rclcpp::spin(node);

  rclcpp::shutdown();
  return EXIT_SUCCESS;
}

File: zivid_samples/scripts/sample_infield_correction.py
#!/usr/bin/env python
from datetime import datetime
import enum
import sys

from rcl_interfaces.msg import ParameterDescriptor
import rclpy
from rclpy.client import Client
from rclpy.executors import ExternalShutdownException
from rclpy.node import Node
from std_srvs.srv import Trigger
from zivid_interfaces.srv import (
    InfieldCorrectionCapture,
    InfieldCorrectionCompute,
    InfieldCorrectionRead,
)


class InfieldCorrectionOperation(enum.Enum):
    Verify = 'verify'
    Correct = 'correct'
    CorrectAndWrite = 'correct_and_write'
    Read = 'read'
    Reset = 'reset'


def capture_status_to_string(response):
    status_map = {
        InfieldCorrectionCapture.Response.STATUS_NOT_SET: 'STATUS_NOT_SET',
        InfieldCorrectionCapture.Response.STATUS_OK: 'STATUS_OK',
        InfieldCorrectionCapture.Response.STATUS_DETECTION_FAILED: 'STATUS_DETECTION_FAILED',
        InfieldCorrectionCapture.Response.STATUS_INVALID_CAPTURE_METHOD: 'STATUS_INVALID_CAPTURE_METHOD',  # noqa: E501
        InfieldCorrectionCapture.Response.STATUS_INVALID_ALIGNMENT: 'STATUS_INVALID_ALIGNMENT',
    }
    if response.status in status_map:
        return status_map[response.status]
    raise RuntimeError(f'Invalid status: {response.status}')


def declare_and_get_parameter_enum(node, name, enum_type):
    name_value_map = {op.value: op for op in enum_type}
    value = (
        node.declare_parameter(name, '', ParameterDescriptor(read_only=True))
        .get_parameter_value()
        .string_value
    )
    if value not in name_value_map:
        valid_values = ', '.join(list(name_value_map.keys()))
        raise RuntimeError(
            f"Invalid value for parameter '{name}': '{value}'. Expected one of: {valid_values}."
        )
    node.get_logger().info(f'Got parameter {name}: {value}')
    return name_value_map[value]


class Sample(Node):

    def __init__(self):
        super().__init__('sample_infield_correction_py')

    def _request_trigger_and_print_response(self, client):
        future = client.call_async(Trigger.Request())
        rclpy.spin_until_future_complete(self, future)
        if not future.done():
            raise RuntimeError(f'Failed to call the service {client.srv_name}')
        result = future.result()
        self.get_logger().info(f'Trigger results ({client.srv_name}):')
        self.get_logger().info(f'  Success: {result.success}')
        message_str = f'"""{result.message}"""' if result.message else ''
        self.get_logger().info(f'  Message: {message_str}')
        return result.success

    def _request_capture_and_print_response(self, client):
        future = client.call_async(InfieldCorrectionCapture.Request())
        rclpy.spin_until_future_complete(self, future)
        if not future.done():
            raise RuntimeError('Failed to call the infield correction capture service')
        result = future.result()
        self.get_logger().info(f'Trigger results ({client.srv_name}):')
        self.get_logger().info(f'  Success: {result.success}')
        message_str = f'"""{result.message}"""' if result.message else ''
        self.get_logger().info(f'  Message: {message_str}')
        self.get_logger().info(f'  Status: {capture_status_to_string(result)}')
        self.get_logger().info(f'  Number of captures: {result.number_of_captures}')
        return result.success

    def _request_infield_correction_compute_and_print_response(
        self, client: Client, name: str
    ):
        future = client.call_async(InfieldCorrectionCompute.Request())
        rclpy.spin_until_future_complete(self, future)
        if not future.done():
            raise RuntimeError(f'Failed to call the infield correction {name} service')
        result = future.result()
        self.get_logger().info(
            f'Infield correction compute results ({client.srv_name}):'
        )
        self.get_logger().info(f'  Success: {result.success}')
        self.get_logger().info(f'  Number of captures: {result.number_of_captures}')
        self.get_logger().info('  Trueness errors:')
        for i, error in enumerate(result.trueness_errors):
            self.get_logger().info(f'  - Capture {i + 1}: {error * 100.0} %')
        self.get_logger().info(
            f'  Average trueness error: {result.average_trueness_error * 100.0} %'
        )
        self.get_logger().info(
            f'  Maximum trueness error: {result.maximum_trueness_error * 100.0} %'
        )
        self.get_logger().info(
            f'  Dimension accuracy: {result.dimension_accuracy * 100.0} %'
        )
        self.get_logger().info(f'  Z min: {result.z_min} m')
        self.get_logger().info(f'  Z max: {result.z_max} m')
        message_str = f'"""{result.message}"""' if result.message else ''
        self.get_logger().info(f'  Message: {message_str}')
        return result.success

    def _request_read_and_print_response(self, client: Client):
        future = client.call_async(InfieldCorrectionRead.Request())
        rclpy.spin_until_future_complete(self, future)
        if not future.done():
            raise RuntimeError('Failed to call the infield correction read service')
        result = future.result()
        time_str = datetime.fromtimestamp(
            result.camera_correction_timestamp.sec
        ).strftime('%c')

        self.get_logger().info(f'Trigger results ({client.srv_name}):')
        self.get_logger().info(f'  Success: {result.success}')
        message_str = f'"""{result.message}"""' if result.message else ''
        self.get_logger().info(f'  Message: {message_str}')
        self.get_logger().info(
            f'  Has camera correction: {result.has_camera_correction}'
        )
        self.get_logger().info(
            f'  Camera correction timestamp: {result.camera_correction_timestamp.sec} ({time_str})'
        )
        return result.success

    def run_operation(self, operation: InfieldCorrectionOperation):
        if operation == InfieldCorrectionOperation.Verify:
            self.get_logger().info('--- Starting infield correction: Verify ---')
            start_client = self.create_client(Trigger, 'infield_correction/start')
            while not start_client.wait_for_service(timeout_sec=3.0):
                self.get_logger().info(
                    f'Service {start_client.srv_name} not available, waiting...'
                )

            capture_client = self.create_client(
                InfieldCorrectionCapture, 'infield_correction/capture'
            )
            compute_client = self.create_client(
                InfieldCorrectionCompute, 'infield_correction/compute'
            )

            if not self._request_trigger_and_print_response(start_client):
                raise RuntimeError('Could not start infield correction, aborting.')
            if not self._request_capture_and_print_response(capture_client):
                raise RuntimeError(
                    'Could not perform an infield correction capture, aborting.'
                )

            self._request_infield_correction_compute_and_print_response(
                compute_client, 'compute'
            )

        elif operation in [
            InfieldCorrectionOperation.Correct,
            InfieldCorrectionOperation.CorrectAndWrite,
        ]:
            self.get_logger().info('--- Starting infield correction ---')
            start_client = self.create_client(Trigger, 'infield_correction/start')
            while not start_client.wait_for_service(timeout_sec=3.0):
                self.get_logger().info(
                    f'Service {start_client.srv_name} not available, waiting...'
                )

            if not self._request_trigger_and_print_response(start_client):
                raise RuntimeError('Could not start infield correction, aborting.')

            capture_client = self.create_client(
                InfieldCorrectionCapture, 'infield_correction/capture'
            )
            compute_client = self.create_client(
                InfieldCorrectionCompute, 'infield_correction/compute'
            )

            wait_seconds = 5
            num_successful_captures_target = 3
            self.get_logger().info(
                f'--- Starting captures, will proceed until {num_successful_captures_target} '
                'captures have completed with a detected calibration board ---'
            )

            sleep_before_capture = False
            num_successful_captures = 0
            while num_successful_captures < num_successful_captures_target:
                if sleep_before_capture:
                    self.get_logger().info(
                        f'--- Waiting for {wait_seconds} seconds'
                        'before taking the next capture ---'
                    )
                    rclpy.spin_until_future_complete(
                        self, rclpy.task.Future(), None, wait_seconds
                    )
                else:
                    sleep_before_capture = True

                success = self._request_capture_and_print_response(capture_client)
                if success:
                    num_successful_captures += 1
                    self._request_infield_correction_compute_and_print_response(
                        compute_client, 'compute'
                    )

            self.get_logger().info('--- Captures complete ---')

            if operation == InfieldCorrectionOperation.CorrectAndWrite:
                self.get_logger().info('--- Writing correction results to camera ---')
                compute_and_write_client = self.create_client(
                    InfieldCorrectionCompute, 'infield_correction/compute_and_write'
                )
                self._request_infield_correction_compute_and_print_response(
                    compute_and_write_client, 'compute and write'
                )

        elif operation == InfieldCorrectionOperation.Read:
            self.get_logger().info('--- Starting infield correction: Read ---')
            read_client = self.create_client(
                InfieldCorrectionRead, 'infield_correction/read'
            )
            while not read_client.wait_for_service(timeout_sec=3.0):
                self.get_logger().info(
                    f'Service {read_client.srv_name} not available, waiting...'
                )
            self._request_read_and_print_response(read_client)

        elif operation == InfieldCorrectionOperation.Reset:
            self.get_logger().info('--- Starting infield correction: Reset ---')
            reset_client = self.create_client(Trigger, 'infield_correction/reset')
            while not reset_client.wait_for_service(timeout_sec=3.0):
                self.get_logger().info(
                    f'Service {reset_client.srv_name} not available, waiting...'
                )
            self._request_trigger_and_print_response(reset_client)

        else:
            raise RuntimeError(f'Unknown operation: {operation}')


def main(args=None):
    rclpy.init(args=args)

    try:
        sample = Sample()
        operation = declare_and_get_parameter_enum(
            sample, 'operation', InfieldCorrectionOperation
        )
        sample.run_operation(operation)
        sample.get_logger().info('Spinning node.. Press Ctrl+C to abort.')
        rclpy.spin(sample)

    except KeyboardInterrupt:
        pass
    except ExternalShutdownException:
        sys.exit(1)


if __name__ == '__main__':
    main()
